vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution_unblinded.cc

```diff
@@ -131,176 +114,405 @@ std::string GenerateTokenPayload(
+        std::move(contribution));
     return;
   }
 
-  MakeContribution(viewing_id, std::move(token_list));
+  PreparePublishers(token_list, std::move(contribution));
 }
 
-void Unblinded::MakeContribution(
-    const std::string& viewing_id,
-    ledger::UnblindedTokenList list) {
-  const auto reconcile = ledger_->GetReconcileById(viewing_id);
-
-  if (reconcile.type == ledger::RewardsType::ONE_TIME_TIP ||
-      reconcile.type == ledger::RewardsType::RECURRING_TIP) {
-    const auto callback = std::bind(&Unblinded::ContributionCompleted,
-        this,
-        _1,
-        viewing_id);
-    SendTokens(
-        reconcile.directions.front().publisher_key,
-        reconcile.type,
-        std::move(list),
-        callback);
+void Unblinded::PreparePublishers(
+    std::vector<ledger::UnblindedToken> list,
+    ledger::ContributionInfoPtr contribution) {
+  if (!contribution) {
     return;
   }
 
-  if (reconcile.type == ledger::RewardsType::AUTO_CONTRIBUTE) {
-    PrepareAutoContribution(viewing_id, std::move(list));
+  if (contribution->type == ledger::RewardsType::AUTO_CONTRIBUTE) {
+    auto publisher_list =
+        PrepareAutoContribution(list, contribution->Clone());
+
+    if (publisher_list.empty()) {
+      ContributionCompleted(
+        ledger::Result::AC_TABLE_EMPTY,
+        std::move(contribution));
+      return;
+    }
+
+    contribution->publishers = std::move(publisher_list);
+
+    ledger_->SaveContributionInfo(
+      contribution->Clone(),
+      std::bind(&Unblinded::OnPrepareAutoContribution,
+          this,
+          _1,
+          contribution->contribution_id));
+    return;
   }
+
+  ProcessTokens(contribution->contribution_id);
 }
 
-bool Unblinded::GetStatisticalVotingWinner(
-    double dart,
-    const ledger::ReconcileDirections& directions,
-    Winners* winners) const {
-  if (!winners) {
-    return false;
+ledger::ContributionPublisherList Unblinded::PrepareAutoContribution(
+    std::vector<ledger::UnblindedToken> list,
+    ledger::ContributionInfoPtr contribution) {
+  if (!contribution || list.size() == 0) {
+    return {};
   }
 
-  double upper = 0.0;
-  for (const auto& item : directions) {
-    upper += item.amount_percent / 100.0;
-    if (upper < dart) {
+  const double total_votes = static_cast<double>(list.size());
+  Winners winners;
+  GetStatisticalVotingWinners(
+      total_votes,
+      contribution->amount,
+      std::move(contribution->publishers),
+      &winners);
+
+  ledger::ContributionPublisherList publisher_list;
+  for (auto & winner : winners) {
+    if (winner.second == 0) {
       continue;
     }
 
-    auto iter = winners->find(item.publisher_key);
+    const std::string publisher_key = winner.first;
+    auto publisher = ledger::ContributionPublisher::New();
+    publisher->contribution_id = contribution->contribution_id;
+    publisher->publisher_key = publisher_key;
+    publisher->total_amount =
+        (winner.second / total_votes) * contribution->amount;
```

Comment:
check on `total_votes == 0`?

Reply:
we check for that on line 350 with `list.size() == 0`